File: lib/presentation/arena/arena_world.dart
import 'dart:math' as math;
import 'dart:ui';
import 'package:flutter/foundation.dart';

import '../../core/engine/vec2.dart';
import '../../core/audio/haptics.dart';
import '../../core/storage/profile_store.dart';
import '../../domain/deities/deity.dart';
import '../../domain/deities/deity_catalog.dart';
import '../../domain/threats/threat.dart';
import '../../domain/combat/parry_system.dart';
import '../../domain/combat/wrath_meter.dart';
import '../../domain/waves/wave_director.dart';

/// A short-lived directional flash drawn where the player parried.
class ParryFlash {
  final double angle;
  final bool perfect;
  double t = 0;
  ParryFlash(this.angle, this.perfect);
}

/// A floating combat number / label.
class FloatText {
  final String text;
  final Offset origin;
  final Color color;
  double t = 0;
  FloatText(this.text, this.origin, this.color);
}

/// The complete arena simulation: state, rules and the per-frame update.
///
/// This is the custom replacement for a Flame `FlameGame`. Rendering is done
/// by [ArenaPainter] which repaints from [frame]; the HUD binds to the
/// individual [ValueNotifier]s below.
class ArenaWorld {
  ArenaWorld({this.tutorial = false}) {
    final p = ProfileStore.instance.profile;
    deity = DeityCatalog.byId(p.deity);
    _parryWindow = tutorial ? 0.40 : p.parryWindow; // forgiving while learning
    _wrathPerParry = p.wrathPerParry;
    _guardCapacity = tutorial ? 99 : p.guardCapacity;
    guard.value = _guardCapacity;
  }

  /// When true the arena runs a gentle scripted tutorial instead of waves.
  final bool tutorial;

  // ── Loadout ────────────────────────────────────────────────────────────
  late final Deity deity;
  late double _parryWindow;
  late double _wrathPerParry;
  late int _guardCapacity;

  // ── Geometry (set once the canvas size is known) ──────────────────────────
  Vec2 centre = Vec2(0, 0);
  double arenaRadius = 360; // spawn ring
  double parryRadius = 132; // visual guard-ring guide
  double coreRadius = 58; // where threats land on the deity

  // ── Live state ─────────────────────────────────────────────────────────────
  final List<Threat> threats = [];
  final List<ParryFlash> flashes = [];
  final List<FloatText> floats = [];
  final WaveDirector director = WaveDirector();
  final WrathMeter wrath = WrathMeter();

  bool isOver = false;
  bool isPaused = false;

  // Run statistics
  int _perfectThisRun = 0;
  int _streak = 0;
  int _maxStreak = 0;
  int _titansThisRun = 0;
  int _repelsThisRun = 0;
  bool _waveDamaged = false; // took a hit during the current wave
  int _lastWave = 1;

  // ── Ultimate transient effects ─────────────────────────────────────────────
  double _slowTimer = 0; // tidal surge
  double _flameTimer = 0; // flame ring
  double ultPulse = 0; // 0..1 visual pulse after firing an ult
  UltimateKind? lastUlt;

  // ── Sprites (decoded by the screen and injected here) ────────────────────────
  List<Image> rockImages = const [];
  List<Image> boltImages = const [];

  // ── Screen shake ─────────────────────────────────────────────────────────────
  double _shake = 0;
  Offset shakeOffset = Offset.zero;
  final math.Random _rng = math.Random();

  // ── Tutorial state ─────────────────────────────────────────────────────────
  final ValueNotifier<String?> tutorialHint = ValueNotifier(null);
  bool tutorialFinished = false;
  VoidCallback? onTutorialDone;
  int _tutPhase = 0;
  double _tutTimer = 0;
  int _tutBlessings = 0;

  // ── HUD notifiers ────────────────────────────────────────────────────────
  final ValueNotifier<int> score = ValueNotifier(0);
  final ValueNotifier<int> wave = ValueNotifier(1);
  final ValueNotifier<int> guard = ValueNotifier(3);
  final ValueNotifier<int> streak = ValueNotifier(0);
  final ValueNotifier<double> wrathValue = ValueNotifier(0);
  final ValueNotifier<bool> wrathReady = ValueNotifier(false);
  final ValueNotifier<String?> banner = ValueNotifier(null);
  final ValueNotifier<int> frame = ValueNotifier(0); // repaint driver
  double _bannerTimer = 0;

  VoidCallback? onGameOver;

  int get essenceEarned => (score.value / 12).floor();

  // ── Setup ───────────────────────────────────────────────────────────────────

  void configure(Size size) {
    centre = Vec2(size.width / 2, size.height * 0.47);
    final shortest = math.min(size.width, size.height);
    parryRadius = shortest * 0.33;
    coreRadius = shortest * 0.15;
    // Spawn just past the farthest visible corner so threats glide in.
    arenaRadius = math.sqrt(size.width * size.width + size.height * size.height) / 2 + 40;
    director.reset();
  }

  // ── Per-frame update ─────────────────────────────────────────────────────────

  void update(double dt) {
    if (isOver || isPaused) return;

    // Banner countdown
    if (_bannerTimer > 0) {
      _bannerTimer -= dt;
      if (_bannerTimer <= 0) banner.value = null;
    }

    // Ult timers
    if (_slowTimer > 0) _slowTimer -= dt;
    if (_flameTimer > 0) _flameTimer -= dt;
    if (ultPulse > 0) ultPulse = (ultPulse - dt * 1.6).clamp(0, 1);

    // Screen shake decay
    if (_shake > 0) {
      _shake = (_shake - dt * 26).clamp(0, 40);
      shakeOffset = Offset(
        (_rng.nextDouble() - 0.5) * _shake,
        (_rng.nextDouble() - 0.5) * _shake,
      );
    } else {
      shakeOffset = Offset.zero;
    }

    wrath.decay(dt);

    // Spawn — scripted tutorial or the wave director.
    if (tutorial) {
      _updateTutorial(dt);
    } else {
      director.update(
        dt,
        arenaRadius: arenaRadius,
        liveThreats: threats.where((t) => t.isAlive).length,
        spawn: threats.add,
      );
      if (director.wave != _lastWave) {
        if (!_waveDamaged) _grantTrial('flawless_wave');
        _waveDamaged = false;
        _lastWave = director.wave;
        wave.value = director.wave;
        _showBanner(director.isTitanWave
            ? '⛰  TITAN WAVE ${director.wave}'
            : 'WAVE ${director.wave}');
        if (director.wave == 10) _grantTrial('wave_10');
        if (director.wave == 20) _grantTrial('wave_20');
      }
    }

    // Flame-ring continuous clear + advance every threat.
    final threatDt = _slowTimer > 0 ? dt * 0.45 : dt;
    for (final t in threats) {
      t.update(t.isAlive ? threatDt : dt);

      if (_flameTimer > 0 && t.isAlive && !t.isBlessing && t.radius <= parryRadius) {
        _repelThreat(t, perfect: false, fromUlt: true);
      }

      if (t.isAlive && t.radius <= coreRadius) {
        _onReachedCore(t);
      }
    }

    // Ricochet: a repelled threat flying outward smashes incoming hostiles it
    // passes through — a satisfying chain reaction worth bonus points.
    _resolveRicochets();

    // Reap finished threats
    threats.removeWhere((t) {
      if (t.isAlive) return false;
      if (t.phase == ThreatPhase.repelled) return t.radius > arenaRadius + 60;
      return t.exitT > 0.4;
    });

    // Flashes / floats
    for (final f in flashes) {
      f.t += dt;
    }
    flashes.removeWhere((f) => f.t > 0.32);
    for (final f in floats) {
      f.t += dt;
    }
    floats.removeWhere((f) => f.t > 0.9);

    frame.value++;
  }

  // ── Input ──────────────────────────────────────────────────────────────────

  void onSwipe(Offset delta) {
    if (isOver || isPaused) return;
    final angle = ParrySystem.swipeAngle(delta);
    if (angle == null) return;

    final outcome = ParrySystem.resolve(
      swipeAngle: angle,
      threats: threats,
      coreRadius: coreRadius,
      parryWindow: _parryWindow,
    );

    flashes.add(ParryFlash(angle, outcome.perfect));

    if (!outcome.connected) {
      _streak = 0;
      streak.value = 0;
      return;
    }

    if (outcome.wastedOnBlessing) {
      // No penalty, but the streak holds; the blessing is simply lost.
      return;
    }

    if (outcome.repelled) {
      _repelThreat(outcome.target!, perfect: outcome.perfect, fromUlt: false);
    } else {
      // Partial hit on a titan.
      Haptics.parry();
      _addScore(6, outcome.target!.positionFrom(centre), AegisAccentless.gold);
    }
  }

  void requestUltimate() {
    if (isOver || isPaused) return;
    if (!wrath.consume()) return;
    wrathReady.value = false;
    wrathValue.value = 0;
    ultPulse = 1;
    lastUlt = deity.ultimate;
    Haptics.surge();
    _grantTrial('ult_first');
    ProfileStore.instance.mutate((p) => p.ultimatesUnleashed += 1);

    switch (deity.ultimate) {
      case UltimateKind.chainLightning:
        for (final t in List<Threat>.from(threats)) {
          if (t.isAlive && !t.isBlessing) {
            _repelThreat(t, perfect: true, fromUlt: true);
          }
        }
        _showBanner('⚡ CHAIN LIGHTNING');
        break;
      case UltimateKind.tidalSurge:
        for (final t in threats) {
          if (t.isAlive && !t.isBlessing) t.radius = arenaRadius;
        }
        _slowTimer = 3.0;
        _showBanner('🌊 TIDAL SURGE');
        break;
      case UltimateKind.soulHarvest:
        _nearestHostiles(3).forEach((t) => _repelThreat(t, perfect: true, fromUlt: true));
        guard.value = _guardCapacity;
        _showBanner('💀 SOUL HARVEST');
        break;
      case UltimateKind.flameRing:
        _flameTimer = 4.0;
        _showBanner('🔥 FLAME RING');
        break;
    }
  }

  // ── Resolution helpers ───────────────────────────────────────────────────────

  void _repelThreat(Threat t, {required bool perfect, required bool fromUlt}) {
    if (!t.isAlive) return;
    t.repel();
    _repelsThisRun += 1;

    final base = switch (t.kind) {
      ThreatKind.boulder => 10,
      ThreatKind.darkBolt => 14,
      ThreatKind.shade => 12,
      ThreatKind.titan => 40,
      _ => 8,
    };
    final mult = 1 + (_streak ~/ 5);
    final pts = perfect ? base * 2 * mult : base * mult;

    if (!fromUlt) {
      _streak += 1;
      _maxStreak = math.max(_maxStreak, _streak);
      streak.value = _streak;
      if (_streak == 25) _grantTrial('streak_25');
    }

    if (perfect) {
      _perfectThisRun += 1;
      if (_perfectThisRun == 10) _grantTrial('perfect_10');
      Haptics.perfect();
      _addShake(6);
    } else if (!fromUlt) {
      Haptics.parry();
    }
    if (t.kind == ThreatKind.titan) _addShake(9);

    if (!fromUlt) wrath.add(perfect ? _wrathPerParry * 1.6 : _wrathPerParry);

    if (t.kind == ThreatKind.titan) {
      _titansThisRun += 1;
      _grantTrial('titan_first');
      ProfileStore.instance.mutate((p) {
        p.titansFelled += 1;
        if (p.titansFelled >= 25) p.earnTrial('titan_slayer');
      });
    }

    _grantTrial('first_parry');
    _addScore(pts, t.positionFrom(centre), perfect ? deity.accent : AegisAccentless.parchment);
    _syncWrath();
  }

  void _onReachedCore(Threat t) {
    if (t.kind == ThreatKind.blessing) {
      t.collect();
      _tutBlessings += 1;
      if (guard.value < _guardCapacity) guard.value += 1;
      _addScore(0, t.positionFrom(centre), AegisAccentless.blessing,
          label: '+GUARD');
      Haptics.parry();
      return;
    }
    if (t.kind == ThreatKind.essenceMote) {
      t.collect();
      score.value += 25;
      _addScore(0, t.positionFrom(centre), AegisAccentless.goldBright,
          label: '+25');
      return;
    }

    // Hostile reached the core.
    t.expire();
    _streak = 0;
    streak.value = 0;

    // In the tutorial nothing can hurt you — it simply respawns next frame.
    if (tutorial) return;

    // Guard break.
    _waveDamaged = true;
    guard.value -= 1;
    Haptics.wound();
    _addShake(8);
    if (guard.value <= 0) {
      _end();
    }
  }

  void _addShake(double amount) {
    _shake = (_shake + amount).clamp(0, 16);
  }

  /// Repelled threats flying outward destroy incoming hostiles they overlap.
  void _resolveRicochets() {
    final outgoing =
        threats.where((t) => t.phase == ThreatPhase.repelled && !t.isBlessing);
    for (final o in outgoing) {
      final op = o.positionFrom(centre);
      for (final t in threats) {
        if (!t.isAlive || t.isBlessing) continue;
        if (op.distanceTo(t.positionFrom(centre)) <= o.drawRadius + t.drawRadius) {
          t.repel();
          _repelsThisRun += 1;
          score.value += 15;
          _addScore(0, t.positionFrom(centre), AegisAccentless.goldBright,
              label: 'COMBO +15');
          _addShake(4);
        }
      }
    }
  }

  // ── Tutorial state machine ────────────────────────────────────────────────────

  void _spawnTutorialBoulder() {
    threats.add(Threat(
      kind: ThreatKind.boulder,
      bearing: _rng.nextDouble() * math.pi * 2,
      radius: arenaRadius,
      speed: 58,
      drawRadius: 34,
      variant: _rng.nextInt(5),
      spin: (_rng.nextDouble() - 0.5) * 2,
    ));
  }

  void _spawnTutorialBlessing() {
    threats.add(Threat(
      kind: ThreatKind.blessing,
      bearing: _rng.nextDouble() * math.pi * 2,
      radius: arenaRadius,
      speed: 64,
      drawRadius: 24,
    ));
  }

  void _updateTutorial(double dt) {
    final live = threats.where((t) => t.isAlive).length;
    switch (_tutPhase) {
      case 0:
        tutorialHint.value = '👉  Swipe toward the ROCK to push it away!';
        if (live == 0 && _repelsThisRun < 1) _spawnTutorialBoulder();
        if (_repelsThisRun >= 1) _tutPhase = 1;
        break;
      case 1:
        tutorialHint.value = '✨  Push when it is CLOSE for a PERFECT!';
        if (live == 0 && _repelsThisRun < 2) _spawnTutorialBoulder();
        if (_repelsThisRun >= 2) _tutPhase = 2;
        break;
      case 2:
        tutorialHint.value = '💚  This is a GIFT — do NOT push! Let it reach you.';
        if (live == 0 && _tutBlessings < 1) _spawnTutorialBlessing();
        if (_tutBlessings >= 1) {
          _tutPhase = 3;
          _tutTimer = 0;
        }
        break;
      default:
        tutorialHint.value = '⚡  Pushes fill your POWER bar — then TAP it for a blast!';
        _tutTimer += dt;
        if (_tutTimer > 1.4) _finishTutorial();
    }
  }

  void _finishTutorial() {
    if (tutorialFinished) return;
    tutorialFinished = true;
    tutorialHint.value = null;
    onTutorialDone?.call();
  }

  List<Threat> _nearestHostiles(int n) {
    final hostiles = threats.where((t) => t.isAlive && !t.isBlessing).toList()
      ..sort((a, b) => a.radius.compareTo(b.radius));
    return hostiles.take(n).toList();
  }

  void _addScore(int pts, Vec2 at, Color color, {String? label}) {
    if (pts != 0) score.value += pts;
    floats.add(FloatText(label ?? '+$pts', Offset(at.x, at.y), color));
  }

  void _syncWrath() {
    wrathValue.value = wrath.value;
    wrathReady.value = wrath.isReady;
  }

  void _showBanner(String text, [double seconds = 1.6]) {
    banner.value = text;
    _bannerTimer = seconds;
  }

  void _grantTrial(String id) {
    // Cheap in-memory guard so we only touch disk the first time a trial is
    // actually earned (this is called on the hot path).
    if (ProfileStore.instance.profile.hasTrial(id)) return;
    ProfileStore.instance.mutate((p) => p.earnTrial(id));
  }

  // ── End of run ────────────────────────────────────────────────────────────────

  Future<void> _end() async {
    if (isOver) return;
    isOver = true;
    Haptics.surge();

    await ProfileStore.instance.mutate((p) {
      p.trialsRun += 1;
      p.essence += essenceEarned;
      p.threatsRepelled += _repelsThisRun;
      if (score.value > p.highScore) p.highScore = score.value;
      if (director.wave > p.bestWave) p.bestWave = director.wave;
      if (_maxStreak > p.bestParryStreak) p.bestParryStreak = _maxStreak;
      p.perfectParries += _perfectThisRun;
      // Pantheon trial if everything is unlocked.
      if (p.unlockedDeities.length >= DeityCatalog.all.length) {
        p.earnTrial('pantheon');
      }
    });

    onGameOver?.call();
  }

  // ── Lifecycle ────────────────────────────────────────────────────────────────

  void reset() {
    final p = ProfileStore.instance.profile;
    _parryWindow = p.parryWindow;
    _wrathPerParry = p.wrathPerParry;
    _guardCapacity = p.guardCapacity;

    threats.clear();
    flashes.clear();
    floats.clear();
    director.reset();
    wrath.reset();

    isOver = false;
    isPaused = false;
    _perfectThisRun = 0;
    _streak = 0;
    _maxStreak = 0;
    _titansThisRun = 0;
    _repelsThisRun = 0;
    _waveDamaged = false;
    _lastWave = 1;
    _slowTimer = 0;
    _flameTimer = 0;
    ultPulse = 0;
    lastUlt = null;
    _shake = 0;
    shakeOffset = Offset.zero;

    score.value = 0;
    wave.value = 1;
    guard.value = _guardCapacity;
    streak.value = 0;
    wrathValue.value = 0;
    wrathReady.value = false;
    banner.value = null;
    _bannerTimer = 0;
  }

  bool get flameActive => _flameTimer > 0;
  bool get slowActive => _slowTimer > 0;
  int get titansThisRun => _titansThisRun;
  int get perfectThisRun => _perfectThisRun;
  int get maxStreak => _maxStreak;

  void dispose() {
    score.dispose();
    wave.dispose();
    guard.dispose();
    streak.dispose();
    wrathValue.dispose();
    wrathReady.dispose();
    banner.dispose();
    tutorialHint.dispose();
    frame.dispose();
  }
}

/// Palette mirror that avoids importing the widgets-flavoured palette into the
/// simulation file (keeps this layer free of Material imports).
class AegisAccentless {
  AegisAccentless._();
  static const Color gold = Color(0xFFE8B84B);
  static const Color goldBright = Color(0xFFFFE27A);
  static const Color parchment = Color(0xFFF3E9D2);
  static const Color blessing = Color(0xFF7DE36B);
}

File: ZeusBoltDash/Presentation/Arena/ArenaWorld.swift
import Foundation
import UIKit

struct FloatText: Identifiable {
    let id = UUID()
    var text: String
    var position: Vec2
    var opacity: Double = 1.0
    var elapsed: Double = 0.0
    let duration: Double = 0.9
}

struct ParryFlash {
    var angle: Double
    var perfect: Bool
    var t: Double = 0
}

final class ArenaWorld: ObservableObject {
    // MARK: - Published state
    @Published private(set) var threats: [Threat] = []
    @Published private(set) var guard_: Int
    @Published private(set) var score: Int = 0
    @Published private(set) var streak: Int = 0
    @Published private(set) var wave: Int = 1
    @Published private(set) var wrathFraction: Double = 0
    @Published private(set) var isOver = false
    @Published private(set) var isPaused = false
    @Published private(set) var floatTexts: [FloatText] = []
    @Published private(set) var parryFlashes: [ParryFlash] = []
    @Published private(set) var ultActive = false
    @Published private(set) var ultKind: UltimateKind = .chainLightning
    @Published private(set) var ultTimeLeft: Double = 0
    @Published private(set) var slowMo = false
    @Published private(set) var banner: String?
    @Published private(set) var tutorialHint: String?
    @Published private(set) var tutorialFinished = false
    private var bannerTime = 0.0

    // MARK: - Config
    let deity: Deity
    let arenaRadius: Double
    let coreRadius: Double
    let parryRadius: Double
    let parryWindow: Double
    let maxGuard: Int
    let isTutorial: Bool
    private let wrathPerParry: Double

    // MARK: - Private
    private let wrath: WrathMeter
    private let waves = WaveDirector()

    private var sessionParries = 0
    private var sessionPerfect = 0
    private var bestStreak = 0
    private var titansFelled = 0
    private var ultimatesUsed = 0
    private var waveDamaged = false

    // Trials earned in-memory during the run; applied to the profile on end.
    private(set) var pendingTrials: Set<String> = []

    // Tutorial state machine
    private var tutPhase = 0
    private var tutTimer = 0.0
    private var tutBlessings = 0

    var essenceEarned: Int { score / 12 }

    init(deity: Deity, parryWindow: Double, maxGuard: Int, wrathPerParry: Double,
         arenaRadius: Double, coreRadius: Double, parryRadius: Double,
         isTutorial: Bool = false) {
        self.deity = deity
        self.arenaRadius = arenaRadius
        self.coreRadius = coreRadius
        self.parryRadius = parryRadius
        // A forgiving window while learning, like the Flutter tutorial.
        self.parryWindow = isTutorial ? 0.40 : parryWindow
        self.maxGuard = isTutorial ? 99 : maxGuard
        self.guard_ = isTutorial ? 99 : maxGuard
        self.isTutorial = isTutorial
        self.wrathPerParry = wrathPerParry
        self.wrath = WrathMeter(wrathPerParry: wrathPerParry)
        self.ultKind = deity.ultimate
    }

    // MARK: - Tick
    func update(dt: Double) {
        guard !isOver, !isPaused else { return }
        let effectiveDt = slowMo ? dt * 0.35 : dt

        if isTutorial {
            updateTutorial(dt: dt)
        } else {
            waves.update(dt: effectiveDt, arenaRadius: arenaRadius, liveThreats: threats.filter { !$0.isRepelled }.count) { [weak self] t in
                self?.threats.append(t)
            }

            if wave != waves.wave {
                // A wave completed: reward a flawless clear, then advance.
                if !waveDamaged { pendingTrials.insert("flawless_wave") }
                waveDamaged = false
                wave = waves.wave
                showBanner(waves.isTitanWave ? "TITAN WAVE \(wave)" : "WAVE \(wave)")
                if wave >= 10 { pendingTrials.insert("wave_10") }
                if wave >= 20 { pendingTrials.insert("wave_20") }
            }
        }

        for t in threats { t.update(dt: effectiveDt) }

        if ultActive && ultKind == .flameRing {
            burnRing()
        }

        checkCoreImpacts()
        checkRicochet()
        cullDeadThreats()

        wrath.update(dt: dt)
        wrathFraction = wrath.fraction

        if ultActive {
            ultTimeLeft -= dt
            if ultTimeLeft <= 0 {
                ultActive = false
                slowMo = false
            }
        }

        if banner != nil {
            bannerTime += dt
            if bannerTime > 1.8 { banner = nil }
        }

        updateFloatTexts(dt: dt)
        updateFlashes(dt: dt)
    }

    private func showBanner(_ text: String) {
        banner = text
        bannerTime = 0
    }

    // MARK: - Input
    func onSwipe(delta: Vec2) {
        guard !isOver, !isPaused else { return }
        guard let angle = ParrySystem.swipeAngle(from: delta) else { return }

        let outcome = ParrySystem.resolve(
            swipeAngle: angle,
            threats: threats,
            coreRadius: coreRadius,
            parryWindow: parryWindow,
            arenaRadius: arenaRadius
        )

        // A flash always appears in the swung direction so the shield arc reads
        // clearly on the ring, even on a miss.
        parryFlashes.append(ParryFlash(angle: angle, perfect: outcome.perfect))

        guard outcome.connected, let t = outcome.target else {
            // Missed swing breaks the streak (mirrors Flutter).
            streak = 0
            return
        }

        sessionParries += 1
        pendingTrials.insert("first_parry")
        if outcome.perfect {
            sessionPerfect += 1
            if sessionPerfect >= 10 { pendingTrials.insert("perfect_10") }
        }

        streak += 1
        if streak > bestStreak { bestStreak = streak }
        if streak >= 25 { pendingTrials.insert("streak_25") }
        if outcome.repelled && t.kind == .titan {
            titansFelled += 1
            pendingTrials.insert("titan_first")
        }
        let streakBonus = min(streak / 5, 5)
        let points = (outcome.perfect ? 20 : 10) + streakBonus * 2
        score += points

        wrath.addParry(perfect: outcome.perfect)

        let pos = t.position
        if outcome.perfect {
            HapticsManager.perfectParry()
            emit(text: "PERFECT +\(points)", at: pos, isGold: true)
        } else {
            HapticsManager.parry()
            emit(text: "+\(points)", at: pos, isGold: false)
        }
    }

    func activateUltimate() {
        guard !isOver, !isPaused, wrath.isFull else { return }
        wrath.consume()
        ultimatesUsed += 1
        pendingTrials.insert("ult_first")
        HapticsManager.ultimateActivate()
        ultActive = true
        showBanner(deity.ultimateName)

        switch deity.ultimate {
        case .chainLightning:
            for t in threats where !t.isRepelled && !t.isPickup { t.repel(arena: arenaRadius) }
            score += 50
            emit(text: "CHAIN LIGHTNING!", at: .zero, isGold: true)
            ultTimeLeft = 0.01

        case .tidalSurge:
            for t in threats where !t.isRepelled { t.repel(arena: arenaRadius) }
            slowMo = true
            ultKind = .tidalSurge
            ultTimeLeft = 3.0
            emit(text: "TIDAL SURGE!", at: .zero, isGold: true)

        case .soulHarvest:
            let nearest = threats
                .filter { !$0.isRepelled && !$0.isPickup }
                .sorted { $0.radius < $1.radius }
                .prefix(3)
            for t in nearest { t.repel(arena: arenaRadius) }
            if guard_ < maxGuard { guard_ += 1 }
            score += 30
            emit(text: "SOUL HARVEST!", at: .zero, isGold: true)
            ultTimeLeft = 0.01

        case .flameRing:
            ultKind = .flameRing
            ultTimeLeft = 4.0
            emit(text: "FLAME RING!", at: .zero, isGold: true)
        }
    }

    func pause() { isPaused = true }
    func resume() { isPaused = false }

    var sessionStats: (parries: Int, perfect: Int, wave: Int, bestStreak: Int, titans: Int, ults: Int) {
        (sessionParries, sessionPerfect, isTutorial ? 1 : waves.wave, bestStreak, titansFelled, ultimatesUsed)
    }

    /// Re-arm the arena for another run (used by "FIGHT AGAIN").
    func reset() {
        threats.removeAll()
        floatTexts.removeAll()
        parryFlashes.removeAll()
        waves.reset()
        wrath.reset()

        isOver = false
        isPaused = false
        ultActive = false
        slowMo = false
        ultTimeLeft = 0
        ultKind = deity.ultimate

        score = 0
        streak = 0
        wave = 1
        guard_ = maxGuard
        wrathFraction = 0
        banner = nil
        bannerTime = 0

        sessionParries = 0
        sessionPerfect = 0
        bestStreak = 0
        titansFelled = 0
        ultimatesUsed = 0
        waveDamaged = false
        pendingTrials.removeAll()
    }

    // MARK: - Tutorial state machine
    private func updateTutorial(dt: Double) {
        let live = threats.filter { !$0.isRepelled }.count
        switch tutPhase {
        case 0:
            tutorialHint = "👉  Swipe toward the ROCK to push it away!"
            if live == 0 && sessionParries < 1 { spawnTutorialBoulder() }
            if sessionParries >= 1 { tutPhase = 1 }
        case 1:
            tutorialHint = "✨  Push when it is CLOSE for a PERFECT!"
            if live == 0 && sessionParries < 2 { spawnTutorialBoulder() }
            if sessionParries >= 2 { tutPhase = 2 }
        case 2:
            tutorialHint = "💚  This is a GIFT — do NOT push! Let it reach you."
            if live == 0 && tutBlessings < 1 { spawnTutorialBlessing() }
            if tutBlessings >= 1 { tutPhase = 3; tutTimer = 0 }
        default:
            tutorialHint = "⚡  Pushes fill your POWER bar — then TAP it for a blast!"
            tutTimer += dt
            if tutTimer > 1.4 { finishTutorial() }
        }
    }

    private func spawnTutorialBoulder() {
        threats.append(Threat(kind: .boulder,
                              bearing: Double.random(in: 0 ..< .pi * 2),
                              radius: arenaRadius, speed: 58))
    }

    private func spawnTutorialBlessing() {
        threats.append(Threat(kind: .blessing,
                              bearing: Double.random(in: 0 ..< .pi * 2),
                              radius: arenaRadius, speed: 64))
    }

    private func finishTutorial() {
        guard !tutorialFinished else { return }
        tutorialFinished = true
        tutorialHint = nil
    }

    /// Apply earned trials + lifetime stats to the store. Call once on game over.
    func commitRun(to store: ProfileStore) {
        for id in pendingTrials { store.earnTrial(id) }
        store.recordRun(score: score, wave: sessionStats.wave,
                        threatsRepelled: sessionParries, perfectParries: sessionPerfect,
                        bestStreak: bestStreak, titansFelled: titansFelled,
                        ultimates: ultimatesUsed)
    }

    // MARK: - Private helpers
    private func checkCoreImpacts() {
        for t in threats where !t.isRepelled {
            if t.radius <= coreRadius {
                if t.isPickup {
                    collectPickup(t)
                } else if isTutorial {
                    // Nothing can hurt you while learning — it simply clears.
                    t.repel(arena: arenaRadius)
                    streak = 0
                } else {
                    t.repel(arena: arenaRadius)
                    waveDamaged = true
                    guard_ -= 1
                    streak = 0
                    HapticsManager.hit()
                    emit(text: "GUARD -1", at: .zero, isGold: false)
                    if guard_ <= 0 {
                        isOver = true
                        HapticsManager.gameOver()
                    }
                }
            }
        }
    }

    private func collectPickup(_ t: Threat) {
        t.repel(arena: arenaRadius)
        HapticsManager.blessingCollect()
        if t.kind == .blessing {
            tutBlessings += 1
            if guard_ < maxGuard { guard_ += 1 }
            score += 15
            emit(text: "+GUARD", at: t.position, isGold: true)
        } else {
            score += 8
            emit(text: "+ESSENCE", at: t.position, isGold: true)
        }
    }

    private func checkRicochet() {
        let repelled = threats.filter { $0.isRepelled }
        let alive = threats.filter { !$0.isRepelled && !$0.isPickup }
        for r in repelled {
            for a in alive {
                if r.position.distance(to: a.position) < 28 {
                    a.repel(arena: arenaRadius)
                    score += 15
                    emit(text: "RICOCHET +15", at: a.position, isGold: true)
                }
            }
        }
    }

    private func burnRing() {
        let burnRadius = coreRadius * 2.8
        for t in threats where !t.isRepelled && !t.isPickup {
            if t.radius <= burnRadius { t.repel(arena: arenaRadius); score += 5 }
        }
    }

    private func cullDeadThreats() {
        threats.removeAll { $0.isRepelled && $0.radius > arenaRadius * 1.3 }
    }

    private func emit(text: String, at pos: Vec2, isGold: Bool) {
        floatTexts.append(FloatText(text: text, position: pos))
    }

    private func updateFloatTexts(dt: Double) {
        for i in floatTexts.indices {
            floatTexts[i].elapsed += dt
            floatTexts[i].position.y -= 28 * dt
            floatTexts[i].opacity = max(0, 1.0 - floatTexts[i].elapsed / floatTexts[i].duration)
        }
        floatTexts.removeAll { $0.elapsed >= $0.duration }
    }

    private func updateFlashes(dt: Double) {
        for i in parryFlashes.indices { parryFlashes[i].t += dt }
        parryFlashes.removeAll { $0.t > 0.32 }
    }
}
